### DISCUSSION TITLE
Emitted Warnings from the compiler should have assigned error identifier.

### DISCUSSION THREAD

**Discussion by rochala** [scala/scala3#19698] [Feature Requests]
As of last year, each dotty Error comes with an assigned identifier. It was added as part of the actionable diagnostics https://github.com/lampepfl/dotty/issues/14904.

This allows us to identify errors in other tools, by comparing the codes. Previously we did this with regex. This discussion is a place where we will discuss adding such identifiers to all warnings we emit.

The motivation behind it is almost the same as with error codes:
- we can identify warnings in other tools such as metals, and provide additional operations and utilities based on that information, e.g. language server protocol specifies [DiagnosticTag](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticTag). This tag is used by the LSP client to fade unused code, or cross deprecated one.
- we can provide actionable diagnostics that will remove unused code (this will allow us to easily remove all unused in a specific file without a need of scalafix)

If you have any opinions, comments, tips or anything related to this initiative, feel free to join the discussion.

This feature is going to be implemented by @ghostbuster91

**Comment by rochala** [comment]
I also have this idea, which I'd like to hear your opinion about:

What if we add `CheckUnused` phase to presentation compiler ?
It would emit warnings (which we would map to hints, or just change the logic directly in phase to report hints instead of warnings when `Mode.is(Interactive)`) on each interactive driver compilation. 

The first concern is how expensive this is, but if the performance lose is insignificant, we could then use those warnings and report them as hints in your client. This will effectively work only on the file that is currently opened, and would work even if you don't have `-Wunused` turned on.

This will properly mark unused and deprecated methods in your IDE, thus enhance developer experience, and while not cluttering their build.
We would obviously add an option to disable this feature.

**Comment by ghostbuster91** [comment]
Here is what I have so far if anyone would like to take a look: https://github.com/lampepfl/dotty/pull/19780

**Comment by som-snytt** [comment]
For the first concern, there is a PR to improve performance of the check.
